src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder.cs

```diff
@@ -285,23 +285,83 @@ public DependencyContext Build(string[] userRuntimeAssemblies = null)
             {
-                var runtimeLibrary = new RuntimeLibrary(
+                var runtimeLibrary = new ModifiableRuntimeLibrary(new RuntimeLibrary(
                     type: "reference",
                     name: GetReferenceLibraryName(directReference),
                     version: directReference.Version,
                     hash: string.Empty,
-                    runtimeAssemblyGroups: new[] { new RuntimeAssetGroup(string.Empty, new[] { CreateRuntimeFile(directReference.FileName, directReference.FullPath) }) },
-                    nativeLibraryGroups: new RuntimeAssetGroup[] { },
+                    runtimeAssemblyGroups: [new RuntimeAssetGroup(string.Empty, [CreateRuntimeFile(directReference.FileName, directReference.FullPath)])],
+                    nativeLibraryGroups: [],
                     resourceAssemblies: CreateResourceAssemblies(directReference.ResourceAssemblies),
-                    dependencies: Enumerable.Empty<Dependency>(),
+                    dependencies: [],
                     path: null,
                     hashPath: null,
                     runtimeStoreManifestName: null,
-                    serviceable: false);
+                    serviceable: false));
 
                 runtimeLibraries.Add(runtimeLibrary);
             }
 
+            /*
+             * We now need to modify runtimeLibraries to eliminate those that don't have any runtime assets. We follow the following steps:
+             * 0. Construct a reverse dependencies list: all runtimeLibraries that depend on this one
+             * 1. If runtimeAssemblyGroups, nativeLibraryGroups, dependencies, and resourceAssemblies are all empty, remove this runtimeLibrary as well as any dependencies on it.
+             * 2. Add all runtimeLibraries to a list of to-be-processed libraries called libraryCandidatesForRemoval
+             * 3. libraryCandidatesForRemoval.Pop() --> if there are no runtimeAssemblyGroups, nativeLibraryGroups, or resourceAssemblies, and either dependencies is empty or all
+             *      dependencies have something else that depends on them, remove it (and from libraryCandidatesForRemoval), adding everything that depends on this to 
+             *      libraryCandidatesForRemoval if it isn't already there
+             * Repeat 3 until libraryCandidatesForRemoval is empty
+             */
+            var references = runtimeLibraries.ToDictionary(lib => lib.Library.Name, lib => lib);
+            foreach (var reference in runtimeLibraries)
+            {
+                foreach (var dependency in reference.Library.Dependencies)
+                {
+                    if (references.TryGetValue(dependency.Name, out var dep))
+                    {
+                        dep.Dependents.Add(reference.Library.Name);
+                    }
+                }
+            }
+
+            var unprocessedReferences = runtimeLibraries.ToHashSet();
+            HashSet<ModifiableRuntimeLibrary> temp = new();
```

Comment:
What's the reason for the separate `temp` HashSet?  Originally I thought it was because you couldn't add to the `unprocessedReference` set while iterating over it, but now I see that you're not actually iterating over it, just popping an item out each time through the loop.  Given that, could you remove the separate `temp` list?

If not, it would probably be a good idea to rename `temp` to something more descriptive and add a comment about why it's used.

Reply:
Good point. That's exactly how I'd been thinking about it, but you're right that it wasn't actually being iterated over, so there's no need for a temp HashSet. Deleted it